Print up to 5 last mock invocations for a wrong oder error

Vampire opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/spockframework/spock/1093)
<!-- Reviewable:end -->

leonard84 commented on spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java:
```diff
@@ -20,17 +20,22 @@
   private final List<IMockInvocation> unmatchedInvocations = new ArrayList<>();
   private int currentRegistrationZone = 0;
   private int currentExecutionZone = 0;
+  private final Semaphore previousInvocationsUpdateSemaphore = new Semaphore(1);
```
I don't think we need that, the MockController is already `synchronized` see org.spockframework.mock.runtime.MockController#handle
`unmatchedInvocations` is unguarded as well, you could also use an `ArrayDeque`

Vampire replied:
Hm, I'll have a look again. I just saw it is called on different threads of course, if your SUT works multi threaded. I wasn't aware there is external synchronization

Vampire commented:
> I don't see any tests for the rendering output

Well, there is none, as for the current output which also is not verified in any way, only that the exception is thrown is validated. :-D

Vampire commented:
I removed the additional synchronization, rewrote the existing `@FailsWith` tests as `EmbeddedSpecification` and added one that checks that the last 5 invocations are printed properly.

leonard84 commented on spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java:
```diff
@@ -41,9 +45,13 @@ public void addInteraction(final IMockInteraction interaction) {
         }
         currentExecutionZone = myRegistrationZone;
+        if (previousInvocations.size() == MAX_PREVIOUS_INVOCATIONS) {
+          previousInvocations.remove();
```
If you use `addFirst` and `removeLast` instead then you get the reverse order out of the box, if we change the name to `previousInvocationsInReverseOrder` then there shouldn't be cause for confusion.

Vampire replied:
as you wish

leonard84 commented on spock-specs/src/test/groovy/org/spockframework/smoke/mock/OrderedInteractions.groovy:
```diff
@@ -98,44 +165,54 @@ class OrderedInteractions extends Specification {
     when:
-    list.add("foo")
-    list.add("bar")
+    runner.runFeatureBody """
```
I don't think that is necessarily better, when we do no further validation on exception using `@FailsWith` is more succinct and it doesn't require that the code is put into a String. While EmbeddedSpecs are powerful, they add a layer of indirection, that is not necessary here.

Vampire replied:
Well, I just did what the documentation says and what seems semantically meaningfuller, which is

> or as a replacement for
> exception conditions in certain corner cases where they cannot be used (like
> specifying the behavior of exception conditions). In all other cases,
> exception conditions are preferable.

But ok, changed it back in the three cases where I don't check the message. :-)

leonard84 replied:
IMO this counts as such a corner case, since you basically have to emulate the whole feature just so you can use the exception condition.

leonard84 commented on spock-core/src/main/java/org/spockframework/mock/WrongInvocationOrderError.java:
```diff
@@ -70,6 +80,19 @@ public String getMessage() {
+    } else {
+      builder.append("Previous invocations in reverse order:");
+      List<IMockInvocation> previousInvocations = new ArrayList<>(this.previousInvocations);
+      Collections.reverse(previousInvocations);
```
This should be redundant, see comment below

leonard84 commented on spock-core/src/main/java/org/spockframework/mock/runtime/InteractionScope.java:
```diff
@@ -20,17 +20,21 @@
 import org.spockframework.util.ExceptionUtil;
 
 import java.util.*;
+import java.util.concurrent.ArrayBlockingQueue;
```
Unused imports

Vampire replied:
Whoops, my bad, fixed.
